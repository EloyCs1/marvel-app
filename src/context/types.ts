import { ReactNode } from "react";

export interface FavoriteContextValue {
  favoriteFilter: boolean;
  toggleFavoriteFilter: () => void;
  favoriteIds: number[];
  toggleFavoriteId: (id: number) => void;
}

export interface FavoriteProviderProps {
  children: ReactNode;
  values: FavoriteContextValue;
}
