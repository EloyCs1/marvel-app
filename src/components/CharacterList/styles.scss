.characterList {
  padding: 48px 48px 48px 48px;

  &__favoritesTitle {
    text-transform: uppercase;
    font-size: 32px;
    font-weight: 700;
    line-height: normal;
  }

  &__cardGrid {
    display: grid;
    grid-gap: 16px;
    grid-template-columns: repeat(9, minmax(170px, 1fr));

    @media (max-width: 1807px) {
      grid-template-columns: repeat(8, minmax(170px, 1fr));
    }

    @media (max-width: 1630px) {
      grid-template-columns: repeat(7, minmax(170px, 1fr));
    }

    @media (max-width: 1430px) {
      grid-template-columns: repeat(6, minmax(170px, 1fr));
    }

    @media (max-width: 1230px) {
      grid-template-columns: repeat(5, minmax(170px, 1fr));
    }

    @media (max-width: 1030px) {
      grid-template-columns: repeat(4, minmax(170px, 1fr));
    }

    @media (max-width: 830px) {
      grid-template-columns: repeat(3, minmax(170px, 1fr));
    }

    @media (max-width: 666px) {
      grid-template-columns: repeat(2, minmax(170px, 1fr));
    }

    @media (max-width: 387px) {
      grid-template-columns: repeat(1, minmax(170px, 1fr));
    }
  }

  @media (max-width: 666px) {
    padding: 24px 16px 0 16px;
  }

  .loader {
    margin-top: 96px;
  }
}
