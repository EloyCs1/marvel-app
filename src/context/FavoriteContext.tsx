import { createContext, useCallback, useContext, useState } from "react";
import { FavoriteContextValue, FavoriteProviderProps } from "./types";

export const defaultFavoriteContextValue: FavoriteContextValue = {
  favoriteFilter: false,
  toggleFavoriteFilter: () => console.warn("toggleFavoriteFilter init"),
  favoriteIds: [],
  toggleFavoriteId: () => console.warn("toggleFavoriteId init"),
};

export const FavoriteContext = createContext(defaultFavoriteContextValue);

export const FavoriteProvider = ({ children, values }: FavoriteProviderProps) => {
  const [favoriteFilter, setFavoriteFilter] = useState(values.favoriteFilter);
  const [favoriteIds, setFavoriteIds] = useState(values.favoriteIds);

  const toggleFavoriteFilter = useCallback((value: boolean) => setFavoriteFilter(value), []);
  const toggleFavoriteId = useCallback(
    (id: number) => {
      const index = favoriteIds.indexOf(id);
      if (index > -1) {
        const ids = [...favoriteIds];
        ids.splice(index, 1);
        setFavoriteIds(ids);
        return;
      }
      setFavoriteIds([...favoriteIds, id]);
    },
    [favoriteIds],
  );

  return (
    <FavoriteContext.Provider value={{ favoriteFilter, toggleFavoriteFilter, favoriteIds, toggleFavoriteId }}>
      {children}
    </FavoriteContext.Provider>
  );
};

export const useFavorites = () => useContext(FavoriteContext);
