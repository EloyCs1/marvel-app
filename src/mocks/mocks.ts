import { AppCharacter } from "../service/types";

export const CHARACTER: AppCharacter.Character = {
  description: "Rick Jones has been Hulk's best bud since day one ",
  id: 1017100,
  image: "http://i.annihil.us/u/prod/marvel/i/mg/3/20/5232158de5b16.jpg",
  name: "A-Bomb (HAS)",
};
