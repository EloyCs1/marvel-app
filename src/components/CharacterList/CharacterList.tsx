import { useIntl } from "react-intl";
import { useFavorites } from "src/context/FavoriteContext";

import useCharacterList from "src/hooks/useCharacterList";
import CharacterCard from "../CharacterCard/CharacterCard";
import SearchInput from "src/components/common/SearchInput/SearchInput";
import Loader from "src/components/common/Loader/Loader";
import ErrorComponent from "../common/ErrorComponent/ErrorComponent";

import "./styles.scss";

const CharacterList = () => {
  const testId = "characterList";
  const { formatMessage } = useIntl();
  const { loading, error, characterList, searchText, onChangeSearchText, tryAgain } = useCharacterList();
  const { favoriteFilter, favoriteIds } = useFavorites();

  const count = favoriteFilter
    ? characterList.filter(({ id }) => favoriteIds.includes(id)).length
    : characterList.length;

  return (
    <div className="characterList">
      {favoriteFilter && (
        <p className="characterList__favoritesTitle" data-testid={`${testId}-favoritesTitle`}>
          {formatMessage({ id: "characterList.favorites.title" })}
        </p>
      )}
      <SearchInput
        testId={`${testId}-searchInput`}
        value={searchText}
        loading={loading}
        placeholder={formatMessage({ id: "characterList.searchInput.placeholder" })}
        helpText={formatMessage(
          { id: "characterList.searchInput.helpText" },
          {
            count,
          },
        )}
        onChange={onChangeSearchText}
      />
      {loading ? (
        <Loader testId={`${testId}-loader`} />
      ) : !error ? (
        <div className="characterList__cardGrid" data-testid={`${testId}-cardGrid`}>
          {characterList.map((character) => (
            <CharacterCard key={character.id} character={character} />
          ))}
        </div>
      ) : (
        <ErrorComponent tryAgain={tryAgain} />
      )}
    </div>
  );
};
export default CharacterList;
