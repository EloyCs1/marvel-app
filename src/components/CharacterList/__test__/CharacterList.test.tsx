import { render } from "@testing-library/react";
import { CHARACTERS } from "src/mocks/mocks";
import { AppWrapper } from "src/__test__/test.utils";
import CharacterList from "../CharacterList";
import { defaultFavoriteContextValue } from "src/context/FavoriteContext";

const mockUseCharacterList = {
  loading: false,
  characterList: CHARACTERS,
  searchText: "",
  onChangeSearchText: jest.fn(),
  error: false,
};

jest.mock("src/hooks/useCharacterList", () => ({
  __esModule: true,
  default: () => mockUseCharacterList,
}));

jest.mock("src/components/CharacterCard/CharacterCard", () => ({
  __esModule: true,
  default: () => <div></div>,
}));

jest.mock("src/components/common/SearchInput/SearchInput", () => ({
  __esModule: true,
  default: () => <div></div>,
}));

describe("CharacterList", () => {
  test("Default render", () => {
    const wrapper = render(
      <AppWrapper>
        <CharacterList />
      </AppWrapper>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("Favorites values render", () => {
    const mockValuesFavoriteProvider = {
      ...defaultFavoriteContextValue,
      favoriteFilter: true,
      favoriteIds: [CHARACTERS[0].id],
      toggleFavoriteId: jest.fn(),
    };

    const wrapper = render(
      <AppWrapper valuesFavoriteProvider={mockValuesFavoriteProvider}>
        <CharacterList />
      </AppWrapper>,
    );
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.getByTestId(`characterList-favoritesTitle`)).toBeInTheDocument();
  });

  test("Error render", () => {
    mockUseCharacterList.error = true;
    const wrapper = render(
      <AppWrapper>
        <CharacterList />
      </AppWrapper>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("Loading render", () => {
    mockUseCharacterList.loading = true;
    const wrapper = render(
      <AppWrapper>
        <CharacterList />
      </AppWrapper>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
