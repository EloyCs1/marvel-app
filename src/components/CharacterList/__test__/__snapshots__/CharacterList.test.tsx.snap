// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList Default render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="characterList"
      >
        <div />
        <div
          class="characterList__cardGrid"
          data-testid="characterList-cardGrid"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="characterList"
    >
      <div />
      <div
        class="characterList__cardGrid"
        data-testid="characterList-cardGrid"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CharacterList Error render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="characterList"
      >
        <div />
        <div
          class="errorComponent"
          data-testid="errorComponent"
        >
          <p
            class="errorComponent__title"
          >
            Error
          </p>
          <p
            class="errorComponent__subTitle"
          >
            Upsss, algo salió mal!!!
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="characterList"
    >
      <div />
      <div
        class="errorComponent"
        data-testid="errorComponent"
      >
        <p
          class="errorComponent__title"
        >
          Error
        </p>
        <p
          class="errorComponent__subTitle"
        >
          Upsss, algo salió mal!!!
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CharacterList Favorites values render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="characterList"
      >
        <p
          class="characterList__favoritesTitle"
          data-testid="characterList-favoritesTitle"
        >
          Favoritos
        </p>
        <div />
        <div
          class="characterList__cardGrid"
          data-testid="characterList-cardGrid"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="characterList"
    >
      <p
        class="characterList__favoritesTitle"
        data-testid="characterList-favoritesTitle"
      >
        Favoritos
      </p>
      <div />
      <div
        class="characterList__cardGrid"
        data-testid="characterList-cardGrid"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CharacterList Loading render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="characterList"
      >
        <div />
        <div
          class="loader"
          data-testid="characterList-loader"
        >
          <span
            class="loader__spinner"
            data-testid="characterList-loader-spinner"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="characterList"
    >
      <div />
      <div
        class="loader"
        data-testid="characterList-loader"
      >
        <span
          class="loader__spinner"
          data-testid="characterList-loader-spinner"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
